File: hydra-haskell/src/main/haskell/Hydra/Dsl/Lib/Logic.hs
--| Phantom-typed term DSL for the hydra.lib.logic library

module Hydra.Dsl.Lib.Logic where

import Hydra.Core
import Hydra.Phantoms
import Hydra.Sources.Libraries
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Dsl.Phantoms as Phantoms


and :: TTerm Bool -> TTerm Bool -> TTerm Bool
and = primitive2 _logic_and

ifElse :: TTerm Bool -> TTerm a -> TTerm a -> TTerm a
ifElse = primitive3 _logic_ifElse

not :: TTerm Bool -> TTerm Bool
not = primitive1 _logic_not

or :: TTerm Bool -> TTerm Bool -> TTerm Bool
or = primitive2 _logic_or

----------------------------------------
-- Helpers which are not primitives

ands :: TTerm [Bool] -> TTerm Bool
ands terms = Phantoms.fold (primitive _logic_and) @@ true @@ terms

ors :: TTerm [Bool] -> TTerm Bool
ors terms = Phantoms.fold (primitive _logic_or) @@ false @@ terms

File: hydra-java/src/main/java/hydra/dsl/meta/lib/Logic.java
package hydra.dsl.meta.lib;

import hydra.core.Name;
import hydra.dsl.meta.Expr;
import hydra.phantoms.TTerm;

import static hydra.dsl.meta.Phantoms.*;

/**
 * Phantom-typed term DSL for the hydra.lib.logic library.
 *
 * <p>Mirrors {@code Hydra.Dsl.Meta.Lib.Logic} (Haskell) and
 * {@code hydra.dsl.meta.lib.logic} (Python).
 *
 * <p>Usage: {@code import static hydra.dsl.meta.lib.Logic.*;}
 */
public interface Logic {

    Name _logic_and = new Name("hydra.lib.logic.and");
    Name _logic_ifElse = new Name("hydra.lib.logic.ifElse");
    Name _logic_not = new Name("hydra.lib.logic.not");
    Name _logic_or = new Name("hydra.lib.logic.or");

    static <R> Expr<R> and(TTerm<?> x, TTerm<?> y) {
        return primitive2(_logic_and, x, y);
    }

    static <R> Expr<R> ifElse(TTerm<?> cond, TTerm<?> then_, TTerm<?> else_) {
        return primitive3(_logic_ifElse, cond, then_, else_);
    }

    static <R> Expr<R> not(TTerm<?> x) {
        return primitive1(_logic_not, x);
    }

    static <R> Expr<R> or(TTerm<?> x, TTerm<?> y) {
        return primitive2(_logic_or, x, y);
    }
}
